### PR TITLE
refactoring: Remove unused includes

### PR DESCRIPTION
**Pull request by practicalswift**
As requested by MarcoFalke in https://github.com/bitcoin/bitcoin/pull/16273#issuecomment-521332089:

This PR removes unused includes.

Please note that in contrast to #16273 I'm limiting the scope to the trivial cases of pure removals (i.e. no includes added) to make reviewing easier.

I'm seeking "Concept ACK":s for this obviously non-urgent minor cleanup.

Rationale:
* Avoids unnecessary re-compiles in case of header changes.
* Makes reasoning about code dependencies easier.
* Reduces compile-time memory usage.
* Reduces compilation time.
* Warm fuzzy feeling of being lean :-)

### CODE REVIEW

**Comment by GChuf** [comment]
I'm all for lean and mean stuff.
Concept ACK

**Comment by jb55** [comment]
Concept ACK

**Comment by emilengler** [comment]
Concept ACK

**Comment by promag** [comment]
Concept ACK.

**Comment by l2a5b1** [comment]
Concept ACK

**Comment by practicalswift** [comment]
@MarcoFalke Could this one get a milestone? I think it should have a concrete milestone if valuable enough to keep open :)  I'm trying to trim my set of open PR:s so if no milestone I'd rather close.

**Comment by promag** [comment]
Running `make` on HEAD 150c289b4034fb2101df61998dd71b11f3d7fae1:
```
     1286.75 real      1170.31 user        99.80 sys
 748662784  maximum resident set size
         0  average shared memory size
         0  average unshared data size
         0  average unshared stack size
  35466407  page reclaims
    168811  page faults
         0  swaps
         0  block input operations
         0  block output operations
         0  messages sent
         0  messages received
      2512  signals received
     39561  voluntary context switches
    377820  involuntary context switches
```

And on HEAD^ e00ecb3d7aaee463643e486ca03c318e192b8058:

```
     1440.06 real      1300.06 user       113.79 sys
 747356160  maximum resident set size
         0  average shared memory size
         0  average unshared data size
         0  average unshared stack size
  35429838  page reclaims
    212723  page faults
         0  swaps
         0  block input operations
         0  block output operations
         0  messages sent
         0  messages received
      2381  signals received
     55950  voluntary context switches
    932017  involuntary context switches
```

(`make clean && ccache -C` before each build).

So I'd ACK if this was already rebased and non draft.

**Comment by practicalswift** [comment]
@promag Thanks a lot for benchmarking! I'll update and rebase as soon as I get clarification from @MarcoFalke regarding the milestone :)

**Comment by practicalswift** [comment]
@MarcoFalke Friendly ping :)

**Comment by laanwj** [comment]
A PR only removing lines is very good. I also think early in the 0.20 cycle is a good time to do this.

So to be clear: this removes unnecessary includes that would otherwise cause the file to be included unnecessary, not includes that still end up being included indirectly through another route?

**Comment by practicalswift** [comment]
@laanwj Thanks for the clarification!

> So to be clear: this removes unnecessary includes that would otherwise cause the file to be included unnecessary, not includes that still end up being included indirectly through another route?

Yes, exactly!

Would you mind adding the 0.20 tag? I'll update the PR :)

**Comment by laanwj** [comment]
Let's get this merged soon then, to not have to keep rebasing, at least there's basically zero risk.

**Comment by practicalswift** [comment]
Rebased again :)

I think this one should be ready for final code review!

**Review by ryanofsky — APPROVED** [review]
Code review ACK 084e17cebd424b8e8ced674bc810eef4e6ee5d3b. PR only removes include lines and it still compiles. In the worst case someone might have to explicitly add an include later for something now included implicitly. But maybe some effort was taken to avoid this, and it wouldn't be a tragedy anyway.

I would encourage @practicalswift to be more transparent about tools and methodologies used, since it's unclear how this PR was generated, how the changes should be maintained, and if there were tools used that might deserve credit.

**Comment by ryanofsky** [comment]
Thanks. I'd be curious to see the script if you want to post it somewhere like a gist.

File: src/node/transaction.cpp [hunk 1]
```diff
@@ -6,7 +6,6 @@
 #include <consensus/validation.h>
 #include <net.h>
 #include <net_processing.h>
-#include <txmempool.h>
```
Comment (MarcoFalke) [inline src/node/transaction.cpp]: This is actually used

**Review by maflcko — COMMENTED** [review]
ACK

File: src/node/transaction.cpp (hunk 1 above)
Comment (practicalswift) [reply src/node/transaction.cpp]: Oh, I'll investigate shortly! Any more details?

**Comment by practicalswift** [comment]
Thanks for merging!

See below for a script run after this merge.

I believe the remaining cases to be of the "non-trivial" type where the header directly included is not needed in itself but one of more of the transitively included headers are.

Consider the `src/bench/rpc_blockchain.cpp` example: we are currently including `validation.h` which **transitively** gives us `PROTOCOL_VERSION`, `CBlock`, `CBlockIndex` and `uint256`. But we don't need anything `validation.h` provides in itself.

Instead of including `validation.h` we should include `version.h` for `PROTOCOL_VERSION`, `primitives/block.h` for `CBlock`, `chain.h` for `CBlockIndex` and `uint256.h` for `uint256`.

Like this:

```diff
diff --git a/src/bench/rpc_blockchain.cpp b/src/bench/rpc_blockchain.cpp
index 2fc6f116a..00cec5ecb 100644
--- a/src/bench/rpc_blockchain.cpp
+++ b/src/bench/rpc_blockchain.cpp
@@ -5,9 +5,12 @@
 #include <bench/bench.h>
 #include <bench/data.h>
 
-#include <validation.h>
+#include <chain.h>
+#include <primitives/block.h>
 #include <streams.h>
 #include <rpc/blockchain.h>
+#include <uint256.h>
+#include <version.h>
 
 #include <univalue.h>
```

```
$ ./find_redundant_includes.sh
src/bench/rpc_blockchain.cpp -- unused include: validation.h
src/bench/verify_script.cpp -- unused include: script/standard.h
src/coins.h -- unused include: core_memusage.h
src/core_read.cpp -- unused include: script/sign.h
src/crypto/hkdf_sha256_32.h -- unused include: crypto/hmac_sha256.h
src/index/txindex.h -- unused include: txdb.h
src/net.h -- unused include: addrdb.h
src/net.h -- unused include: limitedmap.h
src/net.h -- unused include: policy/feerate.h
src/noui.cpp -- unused include: util/system.h
src/psbt.h -- unused include: node/transaction.h
src/psbt.h -- unused include: optional.h
src/psbt.h -- unused include: policy/feerate.h
src/qt/bantablemodel.h -- unused include: net.h
src/qt/receivecoinsdialog.cpp -- unused include: wallet/wallet.h
src/qt/sendcoinsdialog.cpp -- unused include: txmempool.h
src/qt/sendcoinsdialog.cpp -- unused include: wallet/fees.h
src/qt/signverifymessagedialog.cpp -- unused include: wallet/wallet.h
src/qt/walletmodel.h -- unused include: script/standard.h
src/qt/winshutdownmonitor.cpp -- unused include: util/system.h
src/rpc/misc.cpp -- unused include: rpc/blockchain.h
src/rpc/util.h -- unused include: node/transaction.h
src/rpc/util.h -- unused include: script/sign.h
src/script/sign.h -- unused include: streams.h
src/support/events.h -- unused include: ios
src/sync.h -- unused include: thread
src/test/fuzz/FuzzedDataProvider.h -- unused include: utility
src/test/fuzz/eval_script.cpp -- unused include: limits
src/test/key_tests.cpp -- unused include: util/system.h
src/txmempool.h -- unused include: random.h
src/uint256.h -- unused include: assert.h
src/util/system.h -- unused include: util/memory.h
src/util/system.h -- unused include: util/time.h
src/wallet/coinselection.h -- unused include: random.h
src/wallet/crypter.h -- unused include: script/signingprovider.h
src/wallet/test/wallet_tests.cpp -- unused include: rpc/server.h
src/wallet/wallet.cpp -- unused include: util/validation.h
src/wallet/walletdb.h -- unused include: script/sign.h
src/zmq/zmqabstractnotifier.h -- unused include: zmq/zmqconfig.h
src/zmq/zmqconfig.h -- unused include: primitives/transaction.h
src/zmq/zmqconfig.h -- unused include: stdarg.h
```

Anyone who wants to address the above cases too? Then we should be as lean as it gets when it comes to redundant includes :)

If you do: let me know if you find any false positives/false negatives given the list above.

One caveat is that I'm currently skipping analysis for `core_io.h` due to the weird naming mismatch between `core_io.h` ⇄ `core_write.cpp`/`core_read.cpp` :)

**Comment by MarcoFalke** [comment]
Screenshot_2019-10-17 Bitcoin Core bench Timeline(1)

https://codespeed.bitcoinperf.com/timeline/?exe=3%2C4%2C2%2C1%2C5&base=1%2B23&ben=build.make.1.gcc.mem-usage&env=1&revs=10&equid=off&quarts=on&extr=on

Screenshot_2019-10-17 Bitcoin Core bench Timeline

https://codespeed.bitcoinperf.com/timeline/?exe=3%2C4%2C2%2C1%2C5&base=1%2B23&ben=build.make.1.clang.mem-usage&env=1&revs=10&equid=off&quarts=on&extr=on

**Comment by MarcoFalke** [comment]
Looks like this cuts 10% peak memory usage on gcc and 5% on clang:

https://codespeed.bitcoinperf.com/changes/?tre=10&rev=ec3ed5a4487886f1c2a35fda0a3289be7b280248&exe=1&env=1

**Pull request merged**